km/km_cpu_init.c

```diff
@@ -667,10 +671,12 @@ void km_machine_setup(km_machine_init_params_t* params)
             __get_cpuid(entry->function, &entry->eax, &entry->ebx, &entry->ecx, &entry->edx);
             break;
          case 0x0:
-            km_infox(KM_TRACE_KVM, "Setting VendorId to '%s'", cpu_vendor_id);
-            memcpy(&entry->ebx, cpu_vendor_id, sizeof(u_int32_t));
-            memcpy(&entry->edx, cpu_vendor_id + sizeof(u_int32_t), sizeof(u_int32_t));
-            memcpy(&entry->ecx, cpu_vendor_id + 2 * sizeof(u_int32_t), sizeof(u_int32_t));
+            if (set_cpu_vendor_id != 0) {
+               km_infox(KM_TRACE_KVM, "Setting VendorId to '%s'", cpu_vendor_id);
+               memcpy(&entry->ebx, cpu_vendor_id, sizeof(u_int32_t));
+               memcpy(&entry->edx, cpu_vendor_id + sizeof(u_int32_t), sizeof(u_int32_t));
+               memcpy(&entry->ecx, cpu_vendor_id + 2 * sizeof(u_int32_t), sizeof(u_int32_t));
+            }
```

Comment:
how does it work when set_cpu_vendor_id is not set.

Reply:
`entry` is an element from `machine.entries` array which is set by `KVM_GET_SUPPORTED_CPUID` ioctl a little above, so it is already set up. I'm not sure why 0x80000002, 0x80000003 and 0x80000004 are treated special, haven't really dug into it